File: sdk/go/provider_server.go
package gestalt

import (
	"context"
	"errors"
	"fmt"
	"net/http"
	"net/url"

	proto "github.com/valon-technologies/gestalt/server/rpc/protov1/v1"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/status"
	"google.golang.org/protobuf/types/known/emptypb"
)

// ProviderServer adapts a [Provider] implementation to the gRPC
// AppProvider service. Most integration-provider authors should use
// [ServeProvider] instead of constructing this directly.
type ProviderServer struct {
	proto.UnimplementedAppProviderServer
	provider   Provider
	executeFn  func(ctx context.Context, operation string, params map[string]any, token string) (*OperationResult, error)
	sessionCat func() (SessionCatalogProvider, bool)
}

// NewProviderServer adapts provider plus router into the gRPC integration
// surface used by gestaltd.
func NewProviderServer[P any, PP interface {
	*P
	Provider
}](provider PP, router *Router[P]) *ProviderServer {
	return &ProviderServer{
		provider: provider,
		executeFn: func(ctx context.Context, operation string, params map[string]any, token string) (*OperationResult, error) {
			if router == nil {
				return nil, fmt.Errorf("router is nil")
			}
			return router.Execute(ctx, (*P)(provider), operation, params, token)
		},
		sessionCat: func() (SessionCatalogProvider, bool) {
			scp, ok := any(provider).(SessionCatalogProvider)
			return scp, ok
		},
	}
}

func (s *ProviderServer) StartProvider(ctx context.Context, req *proto.StartProviderRequest) (*proto.StartProviderResponse, error) {
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "request is required")
	}
	if err := validateProtocolVersion(req.GetProtocolVersion()); err != nil {
		return nil, err
	}
	config := req.GetConfig().AsMap()
	if config == nil {
		config = map[string]any{}
	}
	if err := s.provider.Configure(ctx, req.GetName(), config); err != nil {
		return nil, status.Errorf(codes.Unknown, "configure provider: %v", err)
	}
	return &proto.StartProviderResponse{
		ProtocolVersion: proto.CurrentProtocolVersion,
	}, nil
}

func (s *ProviderServer) GetMetadata(_ context.Context, _ *emptypb.Empty) (*proto.ProviderMetadata, error) {
	_, ok := s.sessionCat()
	return &proto.ProviderMetadata{
		SupportsSessionCatalog: ok,
		MinProtocolVersion:     proto.CurrentProtocolVersion,
		MaxProtocolVersion:     proto.CurrentProtocolVersion,
	}, nil
}

func (s *ProviderServer) Execute(ctx context.Context, req *proto.ExecuteRequest) (*proto.OperationResult, error) {
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "request is required")
	}
	ctx = withRequestContext(ctx, req.GetContext())
	ctx = WithIdempotencyKey(ctx, req.GetIdempotencyKey())
	if len(req.GetConnectionParams()) > 0 {
		ctx = WithConnectionParams(ctx, req.GetConnectionParams())
	}
	result, err := s.executeFn(ctx, req.GetOperation(), req.GetParams().AsMap(), req.GetToken())
	if err != nil {
		return operationResultProto(operationResultFromError(err)), nil
	}
	if result == nil {
		return operationResultProto(operationResult(http.StatusInternalServerError, nilResultMessage)), nil
	}
	return operationResultProto(result), nil
}

func (s *ProviderServer) ResolveHTTPSubject(ctx context.Context, req *proto.ResolveHTTPSubjectRequest) (resp *proto.ResolveHTTPSubjectResponse, err error) {
	if req == nil {
		return nil, status.Error(codes.InvalidArgument, "request is required")
	}
	ctx = withRequestContext(ctx, req.GetContext())
	defer func() {
		if recovered := recover(); recovered != nil {
			_ = recoveredOperationResult("ResolveHTTPSubject", recovered)
			err = status.Error(codes.Internal, internalErrorMessage)
		}
	}()

	resolver, ok := s.provider.(HTTPSubjectResolver)
	if !ok {
		return &proto.ResolveHTTPSubjectResponse{}, nil
	}

	subject, err := resolver.ResolveHTTPSubject(ctx, httpSubjectRequestFromProto(req.GetRequest()))
	if err != nil {
		var opErr *operationError
		if errors.As(err, &opErr) {
			return &proto.ResolveHTTPSubjectResponse{
				RejectStatus:  int32(opErr.status),
				RejectMessage: opErr.message,
			}, nil
		}
		return nil, providerRPCError("resolve http subject", err)
	}
	if subject == nil {
		return &proto.ResolveHTTPSubjectResponse{}, nil
	}

	return &proto.ResolveHTTPSubjectResponse{
		Subject: &proto.SubjectContext{
			Id:                  subject.ID,
			CredentialSubjectId: subject.CredentialSubjectID,
			Email:               subject.Email,
			DisplayName:         subject.DisplayName,
		},
	}, nil
}

func httpSubjectRequestFromProto(req *proto.HTTPSubjectRequest) HTTPSubjectRequest {
	if req == nil {
		return HTTPSubjectRequest{}
	}
	return HTTPSubjectRequest{
		Binding:         req.GetBinding(),
		Method:          req.GetMethod(),
		Path:            req.GetPath(),
		ContentType:     req.GetContentType(),
		Headers:         httpHeaderFromProto(req.GetHeaders()),
		Query:           urlValuesFromProto(req.GetQuery()),
		Params:          req.GetParams().AsMap(),
		RawBody:         append([]byte(nil), req.GetRawBody()...),
		SecurityScheme:  req.GetSecurityScheme(),
		VerifiedSubject: req.GetVerifiedSubject(),
		VerifiedClaims:  cloneVerifiedClaims(req.GetVerifiedClaims()),
	}
}

func httpHeaderFromProto(values map[string]*proto.StringList) http.Header {
	if len(values) == 0 {
		return nil
	}
	out := make(http.Header, len(values))
	for key, value := range values {
		for _, item := range value.GetValues() {
			out.Add(key, item)
		}
	}
	return out
}

func httpHeaderToProto(values http.Header) map[string]*proto.StringList {
	if len(values) == 0 {
		return nil
	}
	out := make(map[string]*proto.StringList, len(values))
	for key, value := range values {
		out[key] = &proto.StringList{Values: append([]string(nil), value...)}
	}
	return out
}

func urlValuesFromProto(values map[string]*proto.StringList) url.Values {
	if len(values) == 0 {
		return nil
	}
	out := make(url.Values, len(values))
	for key, value := range values {
		out[key] = append([]string(nil), value.GetValues()...)
	}
	return out
}

func cloneVerifiedClaims(values map[string]string) map[string]string {
	if len(values) == 0 {
		return nil
	}
	out := make(map[string]string, len(values))
	for key, value := range values {
		out[key] = value
	}
	return out
}

func (s *ProviderServer) GetSessionCatalog(ctx context.Context, req *proto.GetSessionCatalogRequest) (*proto.GetSessionCatalogResponse, error) {
	scp, ok := s.sessionCat()
	if !ok {
		return nil, status.Error(codes.Unimplemented, "provider does not support session catalogs")
	}
	ctx = withRequestContext(ctx, req.GetContext())
	if len(req.GetConnectionParams()) > 0 {
		ctx = WithConnectionParams(ctx, req.GetConnectionParams())
	}
	cat, err := scp.CatalogForRequest(ctx, req.GetToken())
	if err != nil {
		return nil, status.Errorf(codes.Unknown, "session catalog: %v", err)
	}
	pbCat, err := catalogToProto(cat)
	if err != nil {
		return nil, status.Errorf(codes.InvalidArgument, "session catalog: %v", err)
	}
	return &proto.GetSessionCatalogResponse{Catalog: pbCat}, nil
}

func operationResultProto(result *OperationResult) *proto.OperationResult {
	if result == nil {
		return nil
	}
	return &proto.OperationResult{
		Status:  int32(result.Status),
		Headers: httpHeaderToProto(result.Headers),
		Body:    result.Body,
	}
}

func withRequestContext(ctx context.Context, reqCtx *proto.RequestContext) context.Context {
	if reqCtx == nil {
		return ctx
	}
	ctx = context.WithValue(ctx, requestContextKey{}, reqCtx)
	if subject := reqCtx.GetSubject(); subject != nil {
		ctx = WithSubject(ctx, Subject{
			ID:                  subject.GetId(),
			CredentialSubjectID: subject.GetCredentialSubjectId(),
			Email:               subject.GetEmail(),
			DisplayName:         subject.GetDisplayName(),
			Scopes:              cloneStrings(subject.GetScopes()),
			Permissions:         subjectPermissionsFromProto(subject.GetPermissions()),
		})
	}
	if subject := reqCtx.GetAgentSubject(); subject != nil {
		ctx = WithAgentSubject(ctx, Subject{
			ID:                  subject.GetId(),
			CredentialSubjectID: subject.GetCredentialSubjectId(),
			Email:               subject.GetEmail(),
			DisplayName:         subject.GetDisplayName(),
			Scopes:              cloneStrings(subject.GetScopes()),
			Permissions:         subjectPermissionsFromProto(subject.GetPermissions()),
		})
	}
	if credential := reqCtx.GetCredential(); credential != nil {
		ctx = WithCredential(ctx, Credential{
			Mode:       credential.GetMode(),
			SubjectID:  credential.GetSubjectId(),
			Connection: credential.GetConnection(),
			Instance:   credential.GetInstance(),
		})
	}
	if access := reqCtx.GetAccess(); access != nil {
		ctx = WithAccess(ctx, Access{
			Policy: access.GetPolicy(),
			Role:   access.GetRole(),
		})
	}
	if host := reqCtx.GetHost(); host != nil {
		ctx = WithHostContext(ctx, Host{
			PublicBaseURL: host.GetPublicBaseUrl(),
		})
	}
	if workflow := reqCtx.GetWorkflow(); workflow != nil {
		ctx = WithWorkflowContext(ctx, workflow.AsMap())
	}
	if reqCtx.GetToolRefsSet() {
		ctx = WithToolRefsContext(ctx, agentToolRefsFromProto(reqCtx.GetToolRefs()))
	}
	return ctx
}

type requestContextKey struct{}

func requestContextFromContext(ctx context.Context) *proto.RequestContext {
	reqCtx, _ := ctx.Value(requestContextKey{}).(*proto.RequestContext)
	return reqCtx
}

File: sdk/rust/src/provider_server.rs
use std::collections::BTreeMap;
use std::sync::Arc;

use serde::Serialize;
use serde_json::Value;
use tonic::{Request as GrpcRequest, Response as GrpcResponse, Status};

use crate::agent::{AgentToolRef, agent_tool_ref_from_proto};
use crate::api::{
    Access, Credential, HTTPSubjectRequest, Request, Response, Subject, scope_request_context,
};
use crate::catalog::{catalog_to_proto, object_map};
use crate::env::CURRENT_PROTOCOL_VERSION;
use crate::error::{Error, HTTP_INTERNAL_SERVER_ERROR, INTERNAL_ERROR_MESSAGE};
use crate::generated::v1::app_provider_server::AppProvider;
use crate::generated::v1::{
    ExecuteRequest, GetSessionCatalogRequest, GetSessionCatalogResponse, HttpSubjectRequest,
    OperationResult as ProtoOperationResult, ProviderMetadata, ResolveHttpSubjectRequest,
    ResolveHttpSubjectResponse, StartProviderRequest, StartProviderResponse, SubjectContext,
};
use crate::protocol;
use crate::rpc_status::{require_protocol_version, rpc_status};
use crate::{Provider, Router};

const JSON_CONTENT_TYPE: &str = "application/json";

#[derive(Clone)]
/// gRPC integration-provider server used by the Rust runtime.
pub struct ProviderServer<P> {
    provider: Arc<P>,
    router: Router<P>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
/// Serialized operation result returned by the provider runtime.
pub struct OperationResult {
    /// HTTP-style status code.
    pub status: u16,
    /// HTTP response headers.
    pub headers: BTreeMap<String, Vec<String>>,
    /// Serialized response body bytes.
    pub body: Vec<u8>,
}

impl OperationResult {
    /// Returns whether the status is in the HTTP 2xx range.
    pub fn ok(&self) -> bool {
        (200..300).contains(&self.status)
    }

    /// Returns the raw response body bytes.
    pub fn bytes(&self) -> &[u8] {
        &self.body
    }

    /// Decodes the raw response body as UTF-8 text.
    pub fn text(&self) -> String {
        String::from_utf8_lossy(&self.body).into_owned()
    }

    /// Decodes the raw body as JSON without app invocation envelope handling.
    pub fn json(&self) -> std::result::Result<Value, Box<crate::InvokeError>> {
        crate::app_decode::parse_operation_result_json(&self.body).map_err(|_| {
            Box::new(crate::InvokeError {
                app: String::new(),
                operation: String::new(),
                status: None,
                code: None,
                message: "operation result body is not valid JSON".to_string(),
                body: None,
                raw_body: self.body.clone(),
            })
        })
    }

    /// Returns an invoke error when the status is outside the HTTP 2xx range.
    pub fn require_ok(&self) -> std::result::Result<(), Box<crate::InvokeError>> {
        if self.ok() {
            return Ok(());
        }
        Err(Box::new(crate::InvokeError {
            app: String::new(),
            operation: String::new(),
            status: Some(self.status),
            code: None,
            message: format!("app invoke failed with status {}", self.status),
            body: self.json().ok().map(Box::new),
            raw_body: self.body.clone(),
        }))
    }

    /// Serializes a typed handler response.
    pub fn from_response<T: Serialize>(response: Response<T>) -> Self {
        let status = response.status.unwrap_or(200);
        match serde_json::to_vec(&response.body) {
            Ok(body) => Self {
                status,
                headers: json_response_headers(response.headers),
                body,
            },
            Err(error) => {
                eprintln!("internal error in Gestalt operation response: {error}");
                Self::error(HTTP_INTERNAL_SERVER_ERROR, INTERNAL_ERROR_MESSAGE)
            }
        }
    }

    /// Converts an SDK error into a serialized operation result.
    pub fn from_error(error: Error) -> Self {
        let status = error.status().unwrap_or(HTTP_INTERNAL_SERVER_ERROR);
        if !error.expose_message() {
            eprintln!("internal error in Gestalt operation: {}", error.message());
            return Self::error(HTTP_INTERNAL_SERVER_ERROR, INTERNAL_ERROR_MESSAGE);
        }
        Self::error(status, error.message().to_owned())
    }

    /// Creates a serialized error response.
    pub fn error(status: u16, message: impl Into<String>) -> Self {
        Self {
            status,
            headers: json_response_headers(BTreeMap::new()),
            body: serde_json::to_vec(&serde_json::json!({ "error": message.into() }))
                .unwrap_or_else(|_| br#"{"error":"internal error"}"#.to_vec()),
        }
    }
}

fn json_response_headers(
    mut headers: BTreeMap<String, Vec<String>>,
) -> BTreeMap<String, Vec<String>> {
    if !headers
        .keys()
        .any(|name| name.eq_ignore_ascii_case("content-type"))
    {
        headers.insert(
            "Content-Type".to_owned(),
            vec![JSON_CONTENT_TYPE.to_owned()],
        );
    }
    headers
}

impl<P> ProviderServer<P> {
    /// Creates a server from a provider and operation router.
    pub fn new(provider: Arc<P>, router: Router<P>) -> Self {
        Self { provider, router }
    }
}

#[tonic::async_trait]
impl<P> AppProvider for ProviderServer<P>
where
    P: Provider,
{
    async fn get_metadata(
        &self,
        _request: GrpcRequest<()>,
    ) -> std::result::Result<GrpcResponse<ProviderMetadata>, Status> {
        Ok(GrpcResponse::new(ProviderMetadata {
            supports_session_catalog: self.provider.supports_session_catalog(),
            min_protocol_version: CURRENT_PROTOCOL_VERSION,
            max_protocol_version: CURRENT_PROTOCOL_VERSION,
            ..ProviderMetadata::default()
        }))
    }

    async fn start_provider(
        &self,
        request: GrpcRequest<StartProviderRequest>,
    ) -> std::result::Result<GrpcResponse<StartProviderResponse>, Status> {
        let request = request.into_inner();
        require_protocol_version(request.protocol_version, CURRENT_PROTOCOL_VERSION)?;
        self.provider
            .configure(&request.name, object_map(request.config))
            .await
            .map_err(|error| rpc_status("configure provider", error))?;

        Ok(GrpcResponse::new(StartProviderResponse {
            protocol_version: CURRENT_PROTOCOL_VERSION,
        }))
    }

    async fn execute(
        &self,
        request: GrpcRequest<ExecuteRequest>,
    ) -> std::result::Result<GrpcResponse<ProtoOperationResult>, Status> {
        let request = request.into_inner();
        let context = request.context.clone();
        let result = scope_request_context(
            context,
            self.router.execute(
                Arc::clone(&self.provider),
                &request.operation,
                Value::Object(object_map(request.params)),
                request_context(
                    request.context.as_ref(),
                    request.token,
                    request.connection_params.into_iter().collect(),
                    request.idempotency_key.trim().to_string(),
                ),
            ),
        )
        .await;

        Ok(GrpcResponse::new(ProtoOperationResult {
            status: i32::from(result.status),
            headers: protocol::string_lists_to_proto(result.headers),
            body: result.body,
        }))
    }

    async fn get_session_catalog(
        &self,
        request: GrpcRequest<GetSessionCatalogRequest>,
    ) -> std::result::Result<GrpcResponse<GetSessionCatalogResponse>, Status> {
        if !self.provider.supports_session_catalog() {
            return Err(Status::unimplemented(
                "provider does not support session catalogs",
            ));
        }

        let request = request.into_inner();
        let context = request.context.clone();
        let request = request_context(
            request.context.as_ref(),
            request.token,
            request.connection_params.into_iter().collect(),
            String::new(),
        );
        let catalog = scope_request_context(context, self.provider.catalog_for_request(&request))
            .await
            .map_err(|error| rpc_status("session catalog", error))?;

        Ok(GrpcResponse::new(GetSessionCatalogResponse {
            catalog: catalog.as_ref().map(catalog_to_proto),
        }))
    }

    async fn resolve_http_subject(
        &self,
        request: GrpcRequest<ResolveHttpSubjectRequest>,
    ) -> std::result::Result<GrpcResponse<ResolveHttpSubjectResponse>, Status> {
        let request = request.into_inner();
        let context = request.context.clone();
        let subject = scope_request_context(
            context,
            self.provider.resolve_http_subject(
                http_subject_request(request.request.as_ref()),
                &request_context(
                    request.context.as_ref(),
                    String::new(),
                    Default::default(),
                    String::new(),
                ),
            ),
        )
        .await;

        let subject = match subject {
            Ok(subject) => subject,
            Err(error) if error.status().is_some() && error.expose_message() => {
                return Ok(GrpcResponse::new(ResolveHttpSubjectResponse {
                    reject_status: i32::from(error.status().unwrap_or_default()),
                    reject_message: error.message().to_owned(),
                    ..Default::default()
                }));
            }
            Err(error) => return Err(rpc_status("resolve http subject", error)),
        };

        Ok(GrpcResponse::new(ResolveHttpSubjectResponse {
            subject: subject.map(subject_to_proto),
            ..Default::default()
        }))
    }
}

fn request_context(
    context: Option<&crate::generated::v1::RequestContext>,
    token: String,
    connection_params: std::collections::BTreeMap<String, String>,
    idempotency_key: String,
) -> Request {
    Request {
        token,
        connection_params,
        subject: request_subject(context),
        agent_subject: request_subject_field(context, "agent_subject"),
        credential: request_credential(context),
        access: request_access(context),
        host: request_host(context),
        idempotency_key,
        workflow: request_workflow(context),
        tool_refs: request_tool_refs(context),
        tool_refs_set: context
            .map(|context| context.tool_refs_set)
            .unwrap_or(false),
    }
}

fn request_subject(context: Option<&crate::generated::v1::RequestContext>) -> Subject {
    request_subject_field(context, "subject")
}

fn request_subject_field(
    context: Option<&crate::generated::v1::RequestContext>,
    field_name: &str,
) -> Subject {
    let Some(context) = context else {
        return Subject::default();
    };
    let subject = match field_name {
        "agent_subject" => context.agent_subject.as_ref(),
        _ => context.subject.as_ref(),
    };
    let Some(subject) = subject else {
        return Subject::default();
    };
    Subject {
        id: subject.id.clone(),
        credential_subject_id: subject.credential_subject_id.clone(),
        email: subject.email.clone(),
        display_name: subject.display_name.clone(),
    }
}

fn request_credential(context: Option<&crate::generated::v1::RequestContext>) -> Credential {
    let Some(context) = context else {
        return Credential::default();
    };
    let Some(credential) = context.credential.as_ref() else {
        return Credential::default();
    };
    Credential {
        mode: credential.mode.clone(),
        subject_id: credential.subject_id.clone(),
        connection: credential.connection.clone(),
        instance: credential.instance.clone(),
    }
}

fn request_access(context: Option<&crate::generated::v1::RequestContext>) -> Access {
    let Some(context) = context else {
        return Access::default();
    };
    let Some(access) = context.access.as_ref() else {
        return Access::default();
    };
    Access {
        policy: access.policy.clone(),
        role: access.role.clone(),
    }
}

fn request_host(context: Option<&crate::generated::v1::RequestContext>) -> crate::Host {
    let Some(context) = context else {
        return crate::Host::default();
    };
    let Some(host) = context.host.as_ref() else {
        return crate::Host::default();
    };
    crate::Host {
        public_base_url: host.public_base_url.clone(),
    }
}

fn request_workflow(
    context: Option<&crate::generated::v1::RequestContext>,
) -> serde_json::Map<String, serde_json::Value> {
    let Some(context) = context else {
        return serde_json::Map::new();
    };
    crate::catalog::object_map(context.workflow.clone())
}

fn request_tool_refs(context: Option<&crate::generated::v1::RequestContext>) -> Vec<AgentToolRef> {
    let Some(context) = context else {
        return Vec::new();
    };
    context
        .tool_refs
        .iter()
        .cloned()
        .map(agent_tool_ref_from_proto)
        .collect()
}

fn http_subject_request(request: Option<&HttpSubjectRequest>) -> HTTPSubjectRequest {
    let Some(request) = request else {
        return HTTPSubjectRequest::default();
    };
    HTTPSubjectRequest {
        binding: request.binding.clone(),
        method: request.method.clone(),
        path: request.path.clone(),
        content_type: request.content_type.clone(),
        headers: protocol::string_lists_from_proto(&request.headers),
        query: protocol::string_lists_from_proto(&request.query),
        params: object_map(request.params.clone()),
        raw_body: request.raw_body.clone(),
        security_scheme: request.security_scheme.clone(),
        verified_subject: request.verified_subject.clone(),
        verified_claims: request.verified_claims.clone(),
    }
}

fn subject_to_proto(subject: Subject) -> SubjectContext {
    SubjectContext {
        id: subject.id,
        credential_subject_id: subject.credential_subject_id,
        email: subject.email,
        display_name: subject.display_name,
        ..Default::default()
    }
}
